components/server/ee/src/workspace/workspace-factory.ts

```diff
@@ -63,61 +62,28 @@ export class WorkspaceFactoryEE extends WorkspaceFactory {
             const imageSource = await this.imageSourceProvider.getImageSource(ctx, user, context.actual, config);
 
             // Walk back the commit history to find suitable parent prebuild to start an incremental prebuild on.
+            // TODO this doesn't work for multi-repo projects. We should fetch the last few available prebuilds and iterate over them instead.
             let ws;
-            for (const parent of (context.commitHistory || [])) {
-                const parentPrebuild = await this.db.trace({span}).findPrebuiltWorkspaceByCommit(commitContext.repository.cloneUrl, parent);
-                if (!parentPrebuild) {
-                    continue;
+            if (context.commitHistory && context.commitHistory.length > 0) {
+                const recentPrebuilds = await this.db.trace({span}).findPrebuildsWithWorkpace(commitContext.repository.cloneUrl, 20);
+                const match = recentPrebuilds.find(pb => this.isGoodBaseforIncrementalPrebuild(context, config, imageSource, pb.prebuild, pb.workspace));
+                if (match) {
+                    const incrementalPrebuildContext: PrebuiltWorkspaceContext = {
+                        title: `Incremental prebuild of "${commitContext.title}"`,
+                        originalContext: commitContext,
+                        prebuiltWorkspace: match.prebuild,
+                    }
+                    ws = await this.createForPrebuiltWorkspace({span}, user, incrementalPrebuildContext, normalizedContextURL);
+                    // Overwrite the config from the parent prebuild:
+                    //   `createForPrebuiltWorkspace` 1:1 copies the config from the parent prebuild.
+                    //   Above, we've made sure that the parent's prebuild tasks (before/init/prebuild) are still the same as now.
+                    //   However, other non-prebuild config items might be outdated (e.g. any command task, VS Code extension, ...)
+                    //   To fix this, we overwrite the new prebuild's config with the most-recently fetched config.
+                    // See also: https://github.com/gitpod-io/gitpod/issues/7475
+                    //TODO doing side effects on objects back and forth is complicated and error-prone. We should rather make sure we pass in the config when creating the prebuiltWorkspace.
```

Comment:
nit: please adapt to the TODO comment convention (https://www.notion.so/gitpod/How-we-develop-e8ce7e562478458a9223eb9b797415b2#86e669f8e7df4f1289c9bede0855b5f9)